execute/aggregate.go

```diff
@@ -369,12 +421,53 @@ type aggregateState struct {
 	agg ValueFunc
 }
 
-func (t *simpleAggregateTransformation2) initializeState(chunk table.Chunk, current interface{}) ([]aggregateState, error) {
+// Recyclable represents an aggregate that can reuse state.
+// This is especially useful for transformations like quantile
+// which could potentially consume an inordinate amount of memory
+// if they didn't recycle resources.
+type Recyclable interface {
+	// Recycle should set an aggregate back to it's default state so that it
+	// can be used to calculate another result, and thereby prevent flux from
+	// having to allocate additional memory.
+	Recycle()
+
+	// Drop should decrement the amount of memory set aside for the aggregate.
+	// The value itself will be cleaned up by the garbage collector.
+	Drop()
+}
+
+func (a *aggregateState) Recycle() {
+	if agg, ok := a.agg.(Recyclable); ok {
+		agg.Recycle()
+	}
+}
+
+func (a *aggregateState) Drop() {
+	if agg, ok := a.agg.(Recyclable); ok {
+		agg.Drop()
+	}
+}
+
+type multiState []aggregateState
+
+func (m multiState) Recycle() {
+	for _, state := range m {
+		state.Recycle()
+	}
+}
+
+func (m multiState) Drop() {
+	for _, state := range m {
+		state.Drop()
+	}
+}
+
+func (t *simpleAggregateTransformation2) initializeState(chunk table.Chunk, current interface{}) (multiState, error) {
 	if current != nil {
-		return current.([]aggregateState), nil
+		return current.(multiState), nil
```

Comment:
What are the multiple states being tracked? I assumed there would be one state per table key. Are there more?